tracing/src/main/java/com/palantir/tracing/Tracer.java

```diff
@@ -374,6 +375,42 @@ static String getRequestId(DetachedSpan detachedSpan) {
         throw new SafeIllegalStateException("Unknown span type", SafeArg.of("detachedSpan", detachedSpan));
     }
 
+    @Nullable
+    static <T> T getTraceLocalValue(TraceLocal<T> traceLocal) {
+        Trace maybeCurrentTrace = currentTrace.get();
+        if (maybeCurrentTrace == null) {
+            return null;
+        }
+
+        return (T) maybeCurrentTrace.getTraceState().getTraceLocals().get(traceLocal);
+    }
+
+    @Nullable
+    static <T> T getTraceLocalValue(TraceLocal<T> traceLocal, Supplier<T> initialValue) {
+        Trace maybeCurrentTrace = currentTrace.get();
+        if (maybeCurrentTrace == null) {
+            return null;
+        }
+
+        return (T) maybeCurrentTrace
+                .getTraceState()
+                .getTraceLocals()
+                .computeIfAbsent(traceLocal, _key -> initialValue.get());
+    }
+
+    static <T> void setTraceLocalValue(TraceLocal<T> traceLocal, @Nullable T value) {
+        Trace maybeCurrentTrace = currentTrace.get();
+        if (maybeCurrentTrace == null) {
+            return;
+        }
+
+        if (value != null) {
+            maybeCurrentTrace.getTraceState().getTraceLocals().put(traceLocal, value);
+        } else {
+            maybeCurrentTrace.getTraceState().getTraceLocals().remove(traceLocal);
+        }
```

Comment:
Is it worth pulling out traceLocals map to local since it is on both paths to shrink method bytecode size? Also nit on avoiding negation:

```suggestion
        Map<TraceLocal<?>, Object> traceLocals = maybeCurrentTrace.getTraceState().getTraceLocals();
        if (value == null) {
            traceLocals.remove(traceLocal);
        } else {
            traceLocals.put(traceLocal, value);
        }
```

Reply:
Also `set` shouldn't `remove`, it should only set. Remove should be handled separately, there are subtle performance implications.

Reply:
yeah makes sense